src/support.js

```diff
@@ -84,48 +84,238 @@ var asm2wasmImports = { // special asm2wasm imports
+  function loadLibData() {
 #if WASM
-  var bin;
-  if (lib.buffer) {
-    // we were provided the binary, in a typed array
-    bin = lib;
-  } else {
+    // for wasm, we can use fetch for async, but for fs mode we can only imitate it
+    if (flags.fs) {
+      var libData = flags.fs.readFile(lib, {encoding: 'binary'});
+      if (!(libData instanceof Uint8Array)) {
+        libData = new Uint8Array(lib_data);
+      }
+      return flags.loadAsync ? Promise.resolve(libData) : libData;
+    }
+
+    if (flags.loadAsync) {
+      return fetchBinary(lib);
+    }
     // load the binary synchronously
-    bin = Module['readBinary'](lib);
+    return Module['readBinary'](lib);
+#else
+    // for js we only imitate async for both native & fs modes.
+    var libData;
+    if (flags.fs) {
+      libData = flags.fs.readFile(lib, {encoding: 'utf8'});
+    } else {
+      libData = Module['read'](lib);
+    }
+    return flags.loadAsync ? Promise.resolve(libData) : libData;
+#endif
   }
-  libModule = loadWebAssemblyModule(bin);
+
+  // libModule <- libData
+  function createLibModule(libData) {
+#if WASM
+    return loadWebAssemblyModule(libData, flags)
 #else
-  var src = Module['read'](lib);
-  libModule = eval(src)(
-    alignFunctionTables(),
-    Module
-  );
+    var libModule = eval(libData)(
+      alignFunctionTables(),
+      Module
+    );
+    // load dynamic libraries that this js lib depends on
+    // (wasm loads needed libraries _before_ lib in its own codepath)
+    if (libModule.dynamicLibraries) {
+      if (flags.loadAsync) {
+        return Promise.all(libModule.dynamicLibraries.map(function(dynNeeded) {
+          return loadDynamicLibrary(dynNeeded, flags);
+        })).then(function() {
+          return libModule;
+        });
+      }
+
+      libModule.dynamicLibraries.forEach(function(dynNeeded) {
+        loadDynamicLibrary(dynNeeded, flags);
+      });
+    }
+    return libModule;
 #endif
-  // add symbols into global namespace TODO: weak linking etc.
-  for (var sym in libModule) {
-    if (!Module.hasOwnProperty(sym)) {
-      Module[sym] = libModule[sym];
+  }
+
+  // libModule <- lib
+  function getLibModule() {
+    // lookup preloaded cache first
+    if (Module['preloadedWasm'] !== undefined &&
+        Module['preloadedWasm'][lib] !== undefined) {
+      var libModule = Module['preloadedWasm'][lib];
+      return flags.loadAsync ? Promise.resolve(libModule) : libModule;
+    }
+
+    // module not preloaded - load lib data and create new module from it
+    if (flags.loadAsync) {
+      return loadLibData(lib).then(function(libData) {
+        return createLibModule(libData);
+      });
     }
+
+    return createLibModule(loadLibData(lib));
+  }
+
+  // Module.symbols <- libModule.symbols (flags.global handler)
+  function mergeLibSymbols(libModule) {
+    // add symbols into global namespace TODO: weak linking etc.
+    for (var sym in libModule) {
+      if (!libModule.hasOwnProperty(sym)) {
+        continue;
```

Comment:
this kind of checking just semantically moved, not created here. This is the original code:

https://github.com/kripken/emscripten/blob/a5ef38a9/src/library.js#L1795-L1818

`getOwnPropertyNames` is indeed there https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames but I'm too new to javascript to tell whethere it will be not good here to return the array instead of just iterating.

I suggest to keep the current way for now and do, js modernization, if any in the next patches step by step.

Reply:
OK